src/cbc_sdk/audit_remediation/differential.py

```diff
@@ -0,0 +1,293 @@
+            raise ApiError("Invalid older_run_id")
+        self._older_run_id = older_run_id
+        return self
+
+    def set_device_ids(self, device_ids):
+        """
+        Restricts the query on to the specified devices only.
+
+        Example:
+            >>> run = cb.select(Differential).newer_run_id(newer_run_id).set_device_ids([12345, 56789])
+
+        Args:
+            device_ids (list): List of device id(s)
+
+        Returns:
+            DifferentialQuery: This instance.
+
+        Raises:
+            ApiError: If invalid values are passed in the list.
+        """
+        if not all(isinstance(device_id, int) for device_id in device_ids):
+            raise ApiError("One or more invalid device IDs")
+        self._update_criteria("device_id", device_ids)
+        return self
+
+    def count_only(self, count_only):
+        """
+        Return only count of diff results per device or complete diff metadata result.
+
+        The default value is true, which means only the count will be returned.
+
+        Example:
+            >>> run = cb.select(Differential).newer_run_id(newer_run_id).count_only(True)
+
+        Args:
+            count_only (string): Boolean that indicates whether to return actual metadata
+                or return just the count of differances
+
+        Returns:
+            DifferentialQuery: This instance.
+
+        Raises:
+            ApiError: If invalid values are passed in the list.
+        """
+        if str(count_only).lower() not in ["true", "false"]:
+            raise ApiError("Invalid boolean value for count_only")
+        if str(count_only).lower() == "false":
+            self._count_only = False
+        return self
+
+    def _build_request(self):
+        """
+        Creates the request body for an API call.
+
+        Returns:
+            dict: The complete request body.
+        """
+        request = {"newer_run_id": self._newer_run_id}
+
+        if self._older_run_id:
+            request["older_run_id"] = self._older_run_id
+        if self._criteria:
+            request["criteria"] = self._criteria
+        if self._count_only is False:
+            request["count_only"] = False
+        return request
+
+    def _build_url(self, tail_end=None):
+        """
+        Creates the URL to be used for an API call.
+
+        Args:
+            tail_end (str): String to be appended to the end of the generated URL.
+
+        Returns:
+            str: The complete URL.
+        """
+        url = self._doc_class.urlobject.format(self._cb.credentials.org_key)
+        if tail_end is not None:
+            url += tail_end
+        return url
+
+    def _count(self):
+        """
+        Returns the number of results from the run of this query.
+
+        Returns:
+            int: The number of results from the run of this query.
+        """
+        return self._total_results
+
+    def _perform_query(self, from_row=None, max_rows=None):
```

Comment:
We should implement a helper class which makes it clearer to get the synchronous results instead of relying on `query[0]` given at the moment we only get one result back. We could use `def submit()` as we used a similar experience for Live Query Runs https://carbon-black-cloud-python-sdk.readthedocs.io/en/latest/live-query/#start-a-query-run

Reply:
I used the _perform_query to maintain a consistent experience with the SDK. I think most of the queries return iterators?
I think the submit() will add more complexity for the user - they will have to read the docs to perform the query, where if they receive an iterator, it should be straight-forward for a python user what they need to do with it.
If this is a strong preference, I'll change it do use the `def submit()`

Reply:
I'm stuck on the developer experience given the iterator of the query will only contain one result which is the differential object that contains these properties
```
"newer_run_id": "qpopjb82whlmthlo0x1wrwcnoyxmrueu",
"newer_run_create_time": "2022-05-11T10:49:36.190Z",
"older_run_id": "kibloccplynombvigcgtu2et2zayhzal",
"older_run_create_time": "2022-05-11T10:47:48.952Z",
"diff_processed_time": 0.072,
"newer_run_not_responded_devices": [],
"older_run_not_responded_devices": [],
```

What the developer will likely focus on is the `diff_results` which we can't directly return because the above properties may still be useful.

We already have the following pattern for LiveQuery which uses `.submit()` instead of the query iterator mechanism
```
>>> query_string = "SELECT filename, path FROM file WHERE path = 'C:\\Windows\\Temp\\dbutil_2_3.sys\\' OR path LIKE 'C:\\Users\\%\\AppData\\Local\\Temp\\dbutil_2_3.sys';"
>>> query_object = api.select(Run).where(sql=query_string)
>>> run = query_object.submit()
```

With submit we can return the object directly and avoid needing to return the results list plus if we wanted we could add an async optional parameter to avoid needing another function for `async_export`

The only drawback I see is if in the future they support a multiple run differential but that seems unlikely at the moment given the performance issues but this should be verified before making the final decision

Reply:
Thanks for the feedback; I added the submit() functionality.

Reply:
Thanks for the changes this looks much cleaner